### ISSUE TITLE
Reproduce the karate club example

### ISSUE THREAD

**Issue by riccardobucco** [tkipf/gcn#64]
Hi Thomas,
I'm trying to reproduce the karate club example of your paper using the code of this repo. I can't understand how the data should be saved to match your load_data function.
For example, what about the features of the training/test nodes? I know that the X matrix should be the Identity one, but what's its size? I would appreciate some help, thanks!

**Comment by tkipf** [comment]
Have a look here:
https://github.com/tkipf/gcn/issues/12

**Comment by riccardobucco** [comment]
Sorry for the trivial question.. Which nodes did you use for training/validating/testing? I'm trying to build the three datasets (and respective masks) and i'm not sure about the steps.
So far I have understood what follows:

- **adj** is the adjacency matrix of the graph, with all the nodes (it doesn't matter if they are part of the train/validation/test dataset). Therefore the dimension is NxN, where N is the total number of the nodes in the three datasets.
- **features** is an NxD matrix, but in this case is an NxN identity matrix (again, with N=total number of nodes in the three dataset)
- **y_train**, **y_val** and **y_test** are NxC matrices (C=number of categories), containing zero vectors if the corresponding element in *train_mask*, *val_mask* and *test_mask* is False, or the onehot encode if the corresponding element in the masks is True

Is it correct?
Now, how can I build the three datasets? I don't know if I'm wrong: in your example you labeled only 4 nodes, but you train using also the others (so the training set has N nodes, 4 of which are labeled). This means that y_train must have only 4 vectors using the onehot encoding, while train_mask must have all the elements set to True?
Then you test against the previously unlabeled nodes (N-4), right? So y_test will have only 4 vectors set to zero and test_mask will have the 4 corresponding values set to False.
And what about the validation set? You don't use it, do you?

Thanks in advance, and sorry for making trivial questions.

**Comment by tkipf** [comment]
You can simply set `y_train=y_val=y_test` and also provide the full adjacency matrix and the full feature matrix (i.e., an identity matrix) during training and testing. It suffices to simply provide `train_mask` with elements set to `True` for all nodes that should receive supervision and `False` otherwise.

**Comment by riccardobucco** [comment]
Thanks. Also, does this approach work only with a fixed graph? I mean, let's say I want to add a single node to the graph: do I have to train everything again in order to predict the category of this new node?

**Comment by tkipf** [comment]
You can make predictions on new nodes if their feature vectors come from a similar underlying distribution. If you take unique one-hot vectors as feature vectors (i.e., an identity matrix), then this is not possible (the prediction would mostly be just random).

**Comment by tkipf** [comment]
Yes and yes!

**Comment by riccardobucco** [comment]
what about the out1 dimension? I can't find its value anywhere

**Comment by tkipf** [comment]
Same

**Comment by AugF** [comment]
Sorry to bother you, thanks for your code! It has given me lots of help. But when I try using the karate dataset, I am still confused for two trivial questions:
1.  could you give me the code about y_train, y_val, y_test,  i still don't know what to do.
2.  how to calculate the loss,  in "out2", because your paper said using cross-entropy loss. So i think the dim should be the same,  but how to update the weight of the layer "output -> 2"?
 (my commits may have some grammer mistakes, sorry for it)

**Comment by Abhinav43** [comment]
@AugF  did you find the solution?

I am also looking for code of y_train, y_val, y_test for karate club.

Thanks in advance !

**Comment by Abhinav43** [comment]
@AugF  did you find the solution?

I am also looking for code of y_train, y_val, y_test for karate club.

Thanks in advance !
